binary_transparency/firmware/cmd/ft_witness/internal/http/witness.go

```diff
@@ -0,0 +1,151 @@
+// limitations under the License.
+
+// Package http contains private implementation details for the FirmwareTransparency witness.
+package http
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/binary_transparency/firmware/api"
+	"github.com/google/trillian-examples/binary_transparency/firmware/internal/client"
+	"github.com/google/trillian-examples/binary_transparency/firmware/internal/verify"
+	"github.com/gorilla/mux"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// WitnessStore is the interface to the  Witness Store, for storage of latest checkpoint
+type WitnessStore interface {
+	// Store puts the checkpoint into Witness Store
+	StoreCP(wcp api.LogCheckpoint) error
+
+	// Retrieve gets the stored checkpoint.
+	// Must return status code NotFound if no such checkpoint exists.
+	RetrieveCP() (api.LogCheckpoint, error)
+}
+
+// Witness is the core state & handler implementation of the FT Witness
+type Witness struct {
+	ws           WitnessStore
+	logURL       string
+	pollInterval time.Duration
+}
+
+// NewWitness creates a new Witness.
+func NewWitness(ws WitnessStore, logURL string, pollInterval time.Duration) *Witness {
+	return &Witness{
+		ws:           ws,
+		logURL:       logURL,
+		pollInterval: pollInterval,
+	}
+}
+
+// getCheckpoint returns a checkpoint which is registered with witness
+func (s *Witness) getCheckpoint(w http.ResponseWriter, r *http.Request) {
+	checkpoint, err := s.ws.RetrieveCP()
+	if err != nil {
+		http.Error(w, err.Error(), httpStatusForErr(err))
+		return
+	}
+
+	js, err := json.Marshal(checkpoint)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(js)
+}
+
+// httpStatusForErr maps status codes to HTTP errors.
+func httpStatusForErr(e error) int {
+	switch status.Code(e) {
+	case codes.OK:
+		return http.StatusOK
+	case codes.NotFound:
+		return http.StatusNotFound
+	default:
+		return http.StatusInternalServerError
+	}
+	// unreachable
+}
+
+// RegisterHandlers registers HTTP handlers for firmware transparency endpoints.
+func (s *Witness) RegisterHandlers(r *mux.Router) {
+	r.HandleFunc(fmt.Sprintf("/%s", api.WitnessGetCheckpoint), s.getCheckpoint).Methods("GET")
+}
+
+// Poll periodically polls the FT log for updating the witness checkpoint.
+// It only returns on error (when it doesn't start its own polling thread)
+func (s *Witness) Poll(ctx context.Context) error {
+	ticker := time.NewTicker(s.pollInterval)
+	ftURL, err := url.Parse(s.logURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse FT log URL: %w", err)
+	}
+	glog.Infof("Polling FT log %q...", ftURL)
+	c := client.ReadonlyClient{LogURL: ftURL}
+	lv := verify.NewLogVerifier()
+	for {
+		wcp, err := s.ws.RetrieveCP()
+		if err != nil {
+			glog.Fatal("Failed to retrieve stored logcheckpoint: %w", err)
```

Comment:
This should return the error and the caller should terminate.

Reply:
Can change this...

Reply:
Not needed in the new scheme of things!